scripts/dts/edtlib.py

```diff
@@ -141,6 +144,58 @@ def chosen_node(self, name):
         # to_path() checks that the node exists
         return self._node2enode[chosen.props[name].to_path()]
 
+    def required_by(self, node):
+        """
+        Returns a list of Nodes that have direct dependencies on 'node'.
+        """
+        return self._graph.required_by(node)
+
+    def depends_on(self, node):
+        """
+        Returns a list of Nodes on which 'node' directly depends.
+        """
+        return self._graph.depends_on(node)
+
+    def scc_order(self):
+        """
+        Returns a list of lists of Nodes where all elements of each list
+        depend on each other, and the Nodes in any list do not depend
+        on any Node in a subsequent list.
+
+        For an acyclic graph each list will be a singleton.  Cycles
+        will be represented by lists with multiple nodes.  Cycles are
+        not expected to be present in devicetree graphs.
+        """
+        return self._graph.scc_order()
+
+    def _define_order(self):
+        # Constructs a graph of dependencies between Node instances,
+        # then calculates a partial order over the dependencies.  The
+        # algorithm supports detecting dependency loops.
+
+        self._graph = Graph()
+
+        for node in self.nodes:
+            # A Node always depends on its parent.
+            for child in node.children.values():
+                self._graph.add_edge(child, node)
+
+            # A Node depends on any Nodes present in 'phandle',
+            # 'phandles', or 'phandle-array' property values.
+            for prop in node.props.values():
+                if prop.type == 'phandle':
+                    self._graph.add_edge(node, prop.val)
+                elif prop.type == 'phandles':
+                    for phandle_node in prop.val:
+                        self._graph.add_edge(node, phandle_node)
+                elif prop.type == 'phandle-array':
+                    for cd in prop.val:
+                        self._graph.add_edge(node, cd.controller)
+
+        # Calculate an order that ensures no node is before any node
+        # it depends on.
+        self._graph.scc_order()
```

Comment:
Still need to check for the exception from `_tarjan()` here, and convert it to an `EDTError`.